### ISSUE TITLE
Linux - Dynamically select between Wayland or X11 backends

### ISSUE THREAD

**Issue by JeremyBorys** [glfw/glfw#1655]
While testing the glfw with wayland support enabled at configure/build time, I was hoping that I could switch between X11 or wayland. As far as I could tell, this is a build time decision which means that my application can only ever either use X11 or a Wayland display protocols.

I am hoping that:
 1. there is already an issue for this 
 2. that this might be supported in the future
 3. that it is already available I am missing some environment variable 

I am also not excluding the fact that **my** code could just be broken and **I'm** a terrible programmer.

Basically what I am asking for is:
 * When building with wayland support enabled also build X11 support
 * At runtime determine the display protocol that should be initialised
 * Provide an override environment variable for alternate testing

**Comment by gracicot** [comment]
There was pull request #1338 that started inital work on this.

**Comment by JeremyBorys** [comment]
Thanks @gracicot. Sounds like I can close this since there is some pre-existing work started.

**Comment by ronchaine** [comment]
I also have [a fork](https://github.com/ronchaine/glfw/tree/runtime_backend_choice) that has this functionality, it's not entirely up to date though since I'm not currently using GLFW for anything.

**Comment by linkmauve** [comment]
Oh nice!

Would you like to upstream it maybe?  It would benefit everyone that way. :)

**Comment by ronchaine** [comment]
I wouldn't mind upstreaming this (though it might take a while, I'm currently a bit swamped), but it breaks the ABI and I'm not sure if GLFW is OK with that?

**Comment by tombsar** [comment]
ABI breakage would have to wait until the next major release, and that's probably quite far away. There's a sketchbook for ideas here: https://github.com/glfw/glfw/wiki/GLFW-4.0-sketchbook

**Comment by elmindreda** [comment]
@ronchaine Your branch is really similar to the one I'm working on, but I like your solution for platform data macros better than what I came up with. Mind if I grab bits of your branch (with attribution of course)?

**Comment by ronchaine** [comment]
@elmindreda No, of course not.  Feel free to grab anything that might be helpful!

**Comment by Jacalz** [comment]
Do you have any news regarding the progress of your work for this? @elmindreda

**Comment by ReeceSX** [comment]
Using @ronchaine's fork as a reference, here's some useful copy/pasta

```
sed -i 's/_auWmPlatformInit/_auWmPlatformInitX11/g' *
sed -i 's/_auWmPlatformTerminate/_auWmPlatformTerminateX11/g' *
sed -i 's/_auWmPlatformGetVersionString/_auWmPlatformGetVersionStringX11/g' *
sed -i 's/_auWmPlatformGetCursorPos/_auWmPlatformGetCursorPosX11/g' *
sed -i 's/_auWmPlatformSetCursorPos/_auWmPlatformSetCursorPosX11/g' *
sed -i 's/_auWmPlatformSetCursorMode/_auWmPlatformSetCursorModeX11/g' *
sed -i 's/_auWmPlatformSetWindowMousePassthrough/_auWmPlatformSetWindowMousePassthroughX11/g' *
sed -i 's/_auWmPlatformSetRawMouseMotion/_auWmPlatformSetRawMouseMotionX11/g' *
sed -i 's/_auWmPlatformRawMouseMotionSupported/_auWmPlatformRawMouseMotionSupportedX11/g' *
sed -i 's/_auWmPlatformCreateCursor/_auWmPlatformCreateCursorX11/g' *
sed -i 's/_auWmPlatformCreateStandardCursor/_auWmPlatformCreateStandardCursorX11/g' *
sed -i 's/_auWmPlatformDestroyCursor/_auWmPlatformDestroyCursorX11/g' *
sed -i 's/_auWmPlatformSetCursor/_auWmPlatformSetCursorX11/g' *
sed -i 's/_auWmPlatformGetScancodeName/_auWmPlatformGetScancodeNameX11/g' *
sed -i 's/_auWmPlatformGetKeyScancode/_auWmPlatformGetKeyScancodeX11/g' *
sed -i 's/_auWmPlatformFreeMonitor/_auWmPlatformFreeMonitorX11/g' *
sed -i 's/_auWmPlatformGetMonitorPos/_auWmPlatformGetMonitorPosX11/g' *
sed -i 's/_auWmPlatformGetMonitorContentScale/_auWmPlatformGetMonitorContentScaleX11/g' *
sed -i 's/_auWmPlatformGetMonitorWorkarea/_auWmPlatformGetMonitorWorkareaX11/g' *
sed -i 's/_auWmPlatformGetVideoModes/_auWmPlatformGetVideoModesX11/g' *
sed -i 's/_auWmPlatformGetVideoMode/_auWmPlatformGetVideoModeX11/g' *
sed -i 's/_auWmPlatformGetGammaRamp/_auWmPlatformGetGammaRampX11/g' *
sed -i 's/_auWmPlatformSetGammaRamp/_auWmPlatformSetGammaRampX11/g' *
sed -i 's/_auWmPlatformSetClipboardString/_auWmPlatformSetClipboardStringX11/g' *
sed -i 's/_auWmPlatformGetClipboardString/_auWmPlatformGetClipboardStringX11/g' *
sed -i 's/_auWmPlatformCreateWindow/_auWmPlatformCreateWindowX11/g' *
sed -i 's/_auWmPlatformDestroyWindow/_auWmPlatformDestroyWindowX11/g' *
sed -i 's/_auWmPlatformSetWindowTitle/_auWmPlatformSetWindowTitleX11/g' *
sed -i 's/_auWmPlatformSetWindowIcon/_auWmPlatformSetWindowIconX11/g' *
sed -i 's/_auWmPlatformGetWindowPos/_auWmPlatformGetWindowPosX11/g' *
sed -i 's/_auWmPlatformSetWindowPos/_auWmPlatformSetWindowPosX11/g' *
sed -i 's/_auWmPlatformGetWindowSize/_auWmPlatformGetWindowSizeX11/g' *
sed -i 's/_auWmPlatformSetWindowSize/_auWmPlatformSetWindowSizeX11/g' *
sed -i 's/_auWmPlatformSetWindowSizeLimits/_auWmPlatformSetWindowSizeLimitsX11/g' *
sed -i 's/_auWmPlatformSetWindowAspectRatio/_auWmPlatformSetWindowAspectRatioX11/g' *
sed -i 's/_auWmPlatformGetFramebufferSize/_auWmPlatformGetFramebufferSizeX11/g' *
sed -i 's/_auWmPlatformGetWindowFrameSize/_auWmPlatformGetWindowFrameSizeX11/g' *
sed -i 's/_auWmPlatformGetWindowContentScale/_auWmPlatformGetWindowContentScaleX11/g' *
sed -i 's/_auWmPlatformIconifyWindow/_auWmPlatformIconifyWindowX11/g' *
sed -i 's/_auWmPlatformRestoreWindow/_auWmPlatformRestoreWindowX11/g' *
sed -i 's/_auWmPlatformMaximizeWindow/_auWmPlatformMaximizeWindowX11/g' *
sed -i 's/_auWmPlatformShowWindow/_auWmPlatformShowWindowX11/g' *
sed -i 's/_auWmPlatformHideWindow/_auWmPlatformHideWindowX11/g' *
sed -i 's/_auWmPlatformRequestWindowAttention/_auWmPlatformRequestWindowAttentionX11/g' *
sed -i 's/_auWmPlatformFocusWindow/_auWmPlatformFocusWindowX11/g' *
sed -i 's/_auWmPlatformSetWindowMonitor/_auWmPlatformSetWindowMonitorX11/g' *
sed -i 's/_auWmPlatformWindowFocused/_auWmPlatformWindowFocusedX11/g' *
sed -i 's/_auWmPlatformWindowIconified/_auWmPlatformWindowIconifiedX11/g' *
sed -i 's/_auWmPlatformWindowVisible/_auWmPlatformWindowVisibleX11/g' *
sed -i 's/_auWmPlatformWindowMaximized/_auWmPlatformWindowMaximizedX11/g' *
sed -i 's/_auWmPlatformWindowHovered/_auWmPlatformWindowHoveredX11/g' *
sed -i 's/_auWmPlatformFramebufferTransparent/_auWmPlatformFramebufferTransparentX11/g' *
sed -i 's/_auWmPlatformGetWindowOpacity/_auWmPlatformGetWindowOpacityX11/g' *
sed -i 's/_auWmPlatformSetWindowResizable/_auWmPlatformSetWindowResizableX11/g' *
sed -i 's/_auWmPlatformSetWindowDecorated/_auWmPlatformSetWindowDecoratedX11/g' *
sed -i 's/_auWmPlatformSetWindowFloating/_auWmPlatformSetWindowFloatingX11/g' *
sed -i 's/_auWmPlatformSetWindowOpacity/_auWmPlatformSetWindowOpacityX11/g' *
sed -i 's/_auWmPlatformPollEvents/_auWmPlatformPollEventsX11/g' *
sed -i 's/_auWmPlatformWaitEvents/_auWmPlatformWaitEventsX11/g' *
sed -i 's/_auWmPlatformWaitEventsTimeout/_auWmPlatformWaitEventsTimeoutX11/g' *
sed -i 's/_auWmPlatformPostEmptyEvent/_auWmPlatformPostEmptyEventX11/g' *
sed -i 's/_auWmPlatformGetRequiredInstanceExtensions/_auWmPlatformGetRequiredInstanceExtensionsX11/g' *
sed -i 's/_auWmPlatformGetPhysicalDevicePresentationSupport/_auWmPlatformGetPhysicalDevicePresentationSupportX11/g' *
sed -i 's/_auWmPlatformCreateWindowSurface/_auWmPlatformCreateWindowSurfaceX11/g' *
sed -i 's/_auWmPlatformPollJoystick/_auWmPlatformPollJoystickX11/g' *
sed -i 's/_auWmPlatformUpdateGamepadGUID/_auWmPlatformUpdateGamepadGUIDX11/g' *
```

Ordered `_GLFWplatformfunctions` to play ball with C++, not that this is a C++ library, I just have a C++ fork:
```
static const struct _AUWMplatformfunctions _auWmFunctionsWin32 = {
    ._auWmPlatformInitFunc = _auWmPlatformInitWin32,
    ._auWmPlatformTerminateFunc = _auWmPlatformTerminateWin32,
    ._auWmPlatformGetVersionStringFunc = _auWmPlatformGetVersionStringWin32,
    
    ._auWmPlatformGetCursorPosFunc = _auWmPlatformGetCursorPosWin32,
    ._auWmPlatformSetCursorPosFunc = _auWmPlatformSetCursorPosWin32,
    ._auWmPlatformSetCursorModeFunc = _auWmPlatformSetCursorModeWin32,
    ._auWmPlatformSetWindowMousePassthroughFunc = _auWmPlatformSetWindowMousePassthroughWin32,
    ._auWmPlatformSetRawMouseMotionFunc = _auWmPlatformSetRawMouseMotionWin32,
    ._auWmPlatformRawMouseMotionSupportedFunc = _auWmPlatformRawMouseMotionSupportedWin32,
    ._auWmPlatformCreateCursorFunc = _auWmPlatformCreateCursorWin32,
    ._auWmPlatformCreateStandardCursorFunc = _auWmPlatformCreateStandardCursorWin32,
    ._auWmPlatformDestroyCursorFunc = _auWmPlatformDestroyCursorWin32,
    ._auWmPlatformSetCursorFunc = _auWmPlatformSetCursorWin32,

    ._auWmPlatformGetScancodeNameFunc = _auWmPlatformGetScancodeNameWin32,
    ._auWmPlatformGetKeyScancodeFunc = _auWmPlatformGetKeyScancodeWin32,
    
    ._auWmPlatformFreeMonitorFunc = _auWmPlatformFreeMonitorWin32,
    ._auWmPlatformGetMonitorPosFunc = _auWmPlatformGetMonitorPosWin32,
    ._auWmPlatformGetMonitorContentScaleFunc = _auWmPlatformGetMonitorContentScaleWin32,
    ._auWmPlatformGetMonitorWorkareaFunc = _auWmPlatformGetMonitorWorkareaWin32,
    ._auWmPlatformGetVideoModesFunc = _auWmPlatformGetVideoModesWin32,
    ._auWmPlatformGetVideoModeFunc = _auWmPlatformGetVideoModeWin32,
    ._auWmPlatformGetGammaRampFunc = _auWmPlatformGetGammaRampWin32,
    ._auWmPlatformSetGammaRampFunc = _auWmPlatformSetGammaRampWin32,

    ._auWmPlatformSetClipboardStringFunc = _auWmPlatformSetClipboardStringWin32,
    ._auWmPlatformGetClipboardStringFunc = _auWmPlatformGetClipboardStringWin32,

    ._auWmPlatformCreateWindowFunc = _auWmPlatformCreateWindowWin32,
    ._auWmPlatformDestroyWindowFunc = _auWmPlatformDestroyWindowWin32,
    ._auWmPlatformSetWindowTitleFunc = _auWmPlatformSetWindowTitleWin32,
    ._auWmPlatformSetWindowIconFunc = _auWmPlatformSetWindowIconWin32,
    ._auWmPlatformGetWindowPosFunc = _auWmPlatformGetWindowPosWin32,
    ._auWmPlatformSetWindowPosFunc = _auWmPlatformSetWindowPosWin32,
    ._auWmPlatformGetWindowSizeFunc = _auWmPlatformGetWindowSizeWin32,
    ._auWmPlatformSetWindowSizeFunc = _auWmPlatformSetWindowSizeWin32,
    ._auWmPlatformSetWindowSizeLimitsFunc = _auWmPlatformSetWindowSizeLimitsWin32,
    ._auWmPlatformSetWindowAspectRatioFunc = _auWmPlatformSetWindowAspectRatioWin32,
    ._auWmPlatformGetFramebufferSizeFunc = _auWmPlatformGetFramebufferSizeWin32,
    ._auWmPlatformGetWindowFrameSizeFunc = _auWmPlatformGetWindowFrameSizeWin32,
    ._auWmPlatformGetWindowContentScaleFunc = _auWmPlatformGetWindowContentScaleWin32,
    ._auWmPlatformIconifyWindowFunc = _auWmPlatformIconifyWindowWin32,
    ._auWmPlatformRestoreWindowFunc = _auWmPlatformRestoreWindowWin32,
    ._auWmPlatformMaximizeWindowFunc = _auWmPlatformMaximizeWindowWin32,
    ._auWmPlatformShowWindowFunc = _auWmPlatformShowWindowWin32,
    ._auWmPlatformHideWindowFunc = _auWmPlatformHideWindowWin32,
    ._auWmPlatformRequestWindowAttentionFunc = _auWmPlatformRequestWindowAttentionWin32,
    ._auWmPlatformFocusWindowFunc = _auWmPlatformFocusWindowWin32,
    ._auWmPlatformSetWindowMonitorFunc = _auWmPlatformSetWindowMonitorWin32,
    ._auWmPlatformWindowFocusedFunc = _auWmPlatformWindowFocusedWin32,
    ._auWmPlatformWindowIconifiedFunc = _auWmPlatformWindowIconifiedWin32,
    ._auWmPlatformWindowVisibleFunc =  _auWmPlatformWindowVisibleWin32,
    ._auWmPlatformWindowMaximizedFunc = _auWmPlatformWindowMaximizedWin32,
    ._auWmPlatformWindowHoveredFunc = _auWmPlatformWindowHoveredWin32,
    ._auWmPlatformFramebufferTransparentFunc = _auWmPlatformFramebufferTransparentWin32,
    ._auWmPlatformGetWindowOpacityFunc = _auWmPlatformGetWindowOpacityWin32,
    ._auWmPlatformSetWindowResizableFunc = _auWmPlatformSetWindowResizableWin32,
    ._auWmPlatformSetWindowDecoratedFunc = _auWmPlatformSetWindowDecoratedWin32,
    ._auWmPlatformSetWindowFloatingFunc = _auWmPlatformSetWindowFloatingWin32,
    ._auWmPlatformSetWindowOpacityFunc = _auWmPlatformSetWindowOpacityWin32,

    ._auWmPlatformPollEventsFunc = _auWmPlatformPollEventsWin32,
    ._auWmPlatformWaitEventsFunc = _auWmPlatformWaitEventsWin32,

    ._auWmPlatformWaitEventsTimeoutFunc = _auWmPlatformWaitEventsTimeoutWin32,
    ._auWmPlatformPostEmptyEventFunc = _auWmPlatformPostEmptyEventWin32,
   
    ._auWmPlatformGetRequiredInstanceExtensionsFunc = _auWmPlatformGetRequiredInstanceExtensionsWin32,
    ._auWmPlatformGetPhysicalDevicePresentationSupportFunc = _auWmPlatformGetPhysicalDevicePresentationSupportWin32,
    ._auWmPlatformCreateWindowSurfaceFunc = _auWmPlatformCreateWindowSurfaceWin32,

    ._auWmPlatformPollJoystickFunc = _auWmPlatformPollJoystickMsft,
    ._auWmPlatformUpdateGamepadGUIDFunc = _auWmPlatformUpdateGamepadGUIDMsft
};
````

**Comment by elmindreda** [comment]
Thank you for your patience, everyone. There is now a PR for an initial running implementation of runtime platform selection over in #1958. Feedback is very welcome, there or here.
